options/logger.go

```diff
@@ -78,194 +52,264 @@ func (f LogFormat) MakeFormatter() (send.MessageFormatter, error) {
-	case LogDefault:
-		if l.Options.DefaultPrefix == "" {
-			l.Options.DefaultPrefix = DefaultLogName
-		}
-		sender, err = send.NewNativeLogger(l.Options.DefaultPrefix, l.Options.Level)
-		if err != nil {
-			return nil, err
-		}
-	case LogFile:
-		sender, err = send.NewPlainFileLogger(DefaultLogName, l.Options.FileName, l.Options.Level)
-		if err != nil {
-			return nil, err
-		}
-	case LogInherit:
-		sender = grip.GetSender()
-		if err = sender.SetLevel(l.Options.Level); err != nil {
-			return nil, err
-		}
-	case LogSplunk:
-		if !l.Options.SplunkOptions.Populated() {
-			return nil, errors.New("missing connection info for output type splunk")
+			if len(lc.Config) > 0 {
+				if err := lc.Format.unmarshal(lc.Config, lc.producer); err != nil {
+					return nil, errors.Wrap(err, "problem unmarshalling data")
+				}
+			}
 		}
-		sender, err = send.NewSplunkLogger(DefaultLogName, l.Options.SplunkOptions, l.Options.Level)
-		if err != nil {
-			return nil, err
-		}
-	case LogSumologic:
-		if l.Options.SumoEndpoint == "" {
-			return nil, errors.New("missing endpoint for output type sumologic")
-		}
-		sender, err = send.NewSumo(DefaultLogName, l.Options.SumoEndpoint)
+
+		sender, err := lc.producer.Configure()
 		if err != nil {
 			return nil, err
 		}
-		if err = sender.SetLevel(l.Options.Level); err != nil {
-			return nil, err
-		}
-	case LogInMemory:
-		if l.Options.InMemoryCap <= 0 {
-			return nil, errors.New("invalid inmemory capacity")
-		}
-		sender, err = send.NewInMemorySender(DefaultLogName, l.Options.Level, l.Options.InMemoryCap)
+		lc.sender = sender
+	}
+
+	return lc.sender, nil
+}
+
+func (lc *LoggerConfig) MarshalBSON() ([]byte, error) {
+	if lc.Format == "" {
+		lc.Format = RawLoggerConfigFormatBSON
+	}
+	if lc.Format != RawLoggerConfigFormatBSON {
+		return nil, errors.New("cannot marshal misconfigured bson logger config")
+	}
+
+	if lc.producer != nil {
+		data, err := bson.Marshal(lc.producer)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "problem producing logger config")
 		}
-	case LogBuildloggerV3:
-		sender, err = timber.NewLogger(DefaultLogName, l.Options.Level, &l.Options.BuildloggerV3Options)
+		lc.Config = data
+	}
+
+	return bson.Marshal(struct {
+		Type   string                `bson:"type"`
+		Format RawLoggerConfigFormat `bson:"format"`
+		Config []byte                `bson:"config"`
+	}{Type: lc.Type, Format: lc.Format, Config: lc.Config})
+}
+
+func (lc *LoggerConfig) MarshalJSON() ([]byte, error) {
+	if lc.Format == "" {
+		lc.Format = RawLoggerConfigFormatJSON
+	}
+	if lc.Format != RawLoggerConfigFormatJSON {
+		return nil, errors.New("cannot marshal misconfigured bson logger config")
+	}
+
+	if lc.producer != nil {
+		data, err := json.Marshal(lc.producer)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "problem producing logger config")
 		}
-	default:
-		return nil, errors.New("unknown log type")
+		lc.Config = data
+	}
+
+	return json.Marshal(struct {
```

Comment:
I think duplicating this structure is a bit fraught, and I'm slightly worried that the Config fields will still hold an array of numbers, see earlier comment.